quetz/config.py

```diff
@@ -24,207 +25,276 @@
 PAGINATION_LIMIT = 20
 
 
-class ConfigEntry(NamedTuple):
-    name: str
-    cast: Type
-    default: Any = None
-    required: bool = True
+# class ConfigEntry(NamedTuple):
+#     name: str
+#     cast: Type
+#     default: Any = None
+#     required: bool = True
 
-    def full_name(self, section=""):
-        if section:
-            section += "_"
-        return f"{section}{self.name}"
+#     def full_name(self, section=""):
+#         if section:
+#             section += "_"
+#         return f"{section}{self.name}"
 
-    def env_var(self, section=""):
-        return f"{_env_prefix}{self.full_name(section).upper()}"
+#     def env_var(self, section=""):
+#         return f"{_env_prefix}{self.full_name(section).upper()}"
+
+#     def casted(self, value):
+#         if self.cast is bool:
+#             try:
+#                 value = strtobool(str(value))
+#             except ValueError as e:
+#                 raise ConfigError(f"{self.name}: {e}")
+
+#         return self.cast(value)
+
+
+# class ConfigSection(NamedTuple):
+#     name: str
+#     entries: List[ConfigEntry]
+#     required: bool = True
+
+
+# https://github.com/pydantic/pydantic/issues/1937#issuecomment-695313040
+class DyanamicallyExtendableSetting(BaseSettings):
+    @classmethod
+    def add_fields(cls, **field_definitions: Any):
+        new_fields: Dict[str, ModelField] = {}
+        new_annotations: Dict[str, Optional[type]] = {}
+
+        for f_name, f_def in field_definitions.items():
+            if isinstance(f_def, tuple):
+                try:
+                    f_annotation, f_value = f_def
+                except ValueError as e:
+                    raise Exception(
+                        'field definitions should either be a tuple '
+                        'of (<type>, <default>) or just a'
+                        'default value, unfortunately this means tuples as '
+                        'default values are not allowed'
+                    ) from e
+            else:
+                f_annotation, f_value = None, f_def
+
+            if f_annotation:
+                new_annotations[f_name] = f_annotation
+
+            new_fields[f_name] = ModelField.infer(
+                name=f_name,
+                value=f_value,
+                annotation=f_annotation,
+                class_validators=None,
+                config=cls.__config__,
+            )
+
+        cls.__fields__.update(new_fields)
+        cls.__annotations__.update(new_annotations)
+
+
+class SettingsGeneral(BaseSettings):
+    package_unpack_threads: int = 1
+    frontend_dir: str = ""
+    redirect_http_to_https: bool = False
+
+    class Config:
+        env_prefix = 'general'
+
+
+class SettingsCORS(BaseSettings):
+    allow_origins: list = []
```

Comment:
```suggestion
    allow_origins: list[str] = []
```
I guess these aren't just strings but urls, see another comment on that (in the gitlab settings)